var/spack/repos/builtin/packages/py-petsc4py/package.py

```diff
@@ -50,7 +52,8 @@ class PyPetsc4py(PythonPackage):
     patch("ldshared.patch", when="@:99")
     patch("ldshared-dev.patch", when="@main")
 
-    depends_on("py-cython", type="build", when="@main")
+    depends_on("py-cython@0.29.32:", when="^python@3.11: @3.19:", type="build")
+    depends_on("py-cython@0.24:", when="@3.19:", type="build")
```

Comment:
As is, this says that only main depends on cython.

Reply:
Yes.

So I add this change now - or later with the 3.19 release. [as it won't affect current 3.18 releases..]

Reply:
What I'm saying is that the 3.18 release already contains this constraint in setup.py, but we need to add it to the Spack package too.

Reply:
3.18 tarballs are prebuilt with cython - so there is no dependency on cython here [when spack builds petsc4py].

As noted - it was only required for `@main`

With petsc-3.19 - we are not including the prebuilt cython sources in the tarball - hence need this dependency for 3.19 and newer

So I would think that part of setup.py will be skipped in a spack build with petsc4py-3.18 tarballs

cc: @dalcinl to confirm.

Reply:
We should probably rebuild cython sources even in older versions. This will be done automatically by adding a dependency on cython thanks to #35995. This maximizes the chances that the package will build for all versions of Python.

Reply:
Ok I can add in the dependency for current versions as-well - to force rebuild of cython sources

checking with @dalcinl if there will be any issues with this...

Reply:
@balay Yes, follow @adamjstewart suggestion, it is better to regenerate C sources with Cython to maximize compatibility with newer Python releases.

Reply:
Why don't you just use the latest Cython release 0.29.33 ?

Reply:
This is a lower bound, so if a newer version is available it will use that. We want to keep the range of supported versions as wide as possible in case a different package in the DAG has different constraints.

Reply:
@dalcinl thanks!

the suggested change is pushed

Reply:
But I don't see cython source generated - even though `CYTHON_FORCE_REGEN=1` is set

[spack-build-out.txt](https://github.com/spack/spack/files/11066080/spack-build-out.txt)
[spack-build-env.txt](https://github.com/spack/spack/files/11066081/spack-build-env.txt)

With cython - I would see:
```
running build_src
using Cython 0.29.32
cythonizing 'petsc4py/PETSc.pyx' -> 'petsc4py/PETSc.c'
running build_py
creating build
```

Here I'm seeing:
```
  running build
  running build_src
  running build_py
  creating build
```

Reply:
Hmm, @haampie any ideas?